soccer/planning/planner/SettlePlanner.cpp

```diff
@@ -0,0 +1,177 @@
+        if (!path.empty()) {
+#pragma omp critical
+            {
+                if (path.end_time() + RJ::Seconds{*_bufferTimeBeforeContact} <
+                        contactTime &&
+                    contactTime < interceptTime) {
+                    interceptTime = contactTime;
+                }
+            }
+        }
+    }
+    return interceptTime;
+}
+Trajectory SettlePlanner::intercept(PlanRequest&& request,
+                                    RJ::Time interceptTime) {
+    const Ball& ball = request.context->state.ball;
+    Point ballVel = ball.vel;
+    RobotInstant startInstant = request.start;
+    RotationConstraints rotationConstraints = request.constraints.rot;
+    Point ballPointAtContact = ball.predict(interceptTime).pos;
+    Point robotPointAtContact =
+        ballPointAtContact +
+        ballVel.normalized(Ball_Radius + Robot_MouthRadius);
+    request.static_obstacles.add(std::make_shared<Circle>(
+        ballPointAtContact - ball.vel.norm() * .1,
+        Ball_Radius));  // subtract a bit from ballPointAtContact to prevent a
+                        // collision detection
+    RobotInstant goalInstant{Pose{robotPointAtContact, 0}, {}, RJ::Time{0s}};
+    request.motionCommand = PathTargetCommand{goalInstant};
+    Trajectory path = _pathTargetPlanner.plan(std::move(request));
+    PlanAngles(path, startInstant, AngleFns::faceAngle(ballVel.angle() + M_PI),
+               rotationConstraints);
+    return std::move(path);
+}
+Trajectory SettlePlanner::plan(PlanRequest&& request) {
+    RobotInstant startInstant = request.start;
+    RotationConstraints rotationConstraints = request.constraints.rot;
+    const Ball& ball = request.context->state.ball;
+    Point ballVel = ball.vel;
+    // search for a goal with brute force
+    Pose targetPose = findTargetPose(request);
+    Segment ballLine{ball.pos, targetPose.position()};
+    request.context->debug_drawer.drawLine(ballLine, Qt::white, "Planning");
+    request.context->debug_drawer.drawCircle(
+        targetPose.position(), Robot_MouthRadius, Qt::black, "Planning");
+    request.context->debug_drawer.drawLine(
+        targetPose.position(),
+        targetPose.position() +
+            Point::direction(targetPose.heading()) * 2 * Robot_Radius,
+        Qt::black, "Planning");
+
+    // Shortcut Path: if we are almost done, move directly toward the ball line
+    // This prevents the planner from suddenly changing the target just before
+    // we get there because it thinks we won't make it in time.
+    Point closestPt = ballLine.nearestPoint(startInstant.pose.position());
+    bool inFrontOfBall = ball.vel.dot(closestPt - ball.pos) > 0;
+    bool withinShortcut =
+        startInstant.pose.position().distTo(closestPt) <
+        *_shortcutDist;  // && closestPt.distTo(targetPose.position()) <
+                         // *_shortcutDist;
+    if (inFrontOfBall && withinShortcut) {
```

Comment:
Just to confirm, this also makes sure that if you are very close to the intercept line in terms of distance but far away from the ball, you will not shortcut but instead move towards the ball to intercept sooner? I can describe in pictures if that helps

Reply:
The last two lines of this comment basically https://github.com/RoboJackets/robocup-software/blob/staging/soccer/planning/SettlePathPlanner.cpp#L350

Reply:
I think i got rid of that case because I assume we can't outrun the ball. If we could, I would expect a Collect command to be send instead of Settle. I might be wrong on this though.

Reply:
So this is actually the "opposite" case. If I have a receiver at my goal, and the kicker at their goal. During the pass, assume the ball goes directly to the receiver. In this case, the receiver will be in front of the ball and within shortcut range, so they will sit there at the goal line waiting for the ball. But the ball must travel the entire field. In most every case, there is a better intercept position closer to the half line that will allow you to intercept the ball in less time. The optimal behavior in my head was to cheat forward and intercept at the half line instead of waiting at the goal line for the ball to arrive. The last two lines in the comment refer to this situation. We only want to shortcut if the target point is at or behind us. (Behind refers to intercepting at a later point in time)

Reply:
Yeah, It's a weird case, but I saw it a lot with passes